### PR TITLE
Exclude Top Choices (XTC): Added XTC sampler for ooba

### PR DESCRIPTION
**Pull request by vitorfdl**
Add XTC parameter support for OOBA based on p-e-w contribution: https://github.com/oobabooga/text-generation-webui/pull/6335

Not sure if I've done fully correctly, Sillytavern code is kinda crazy. But it worked for me.

## Checklist:

- [X ] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -1291,6 +1291,26 @@ <h4 class="wide100p textAlignCenter">
                                             </div>
                                         </div>
                                     </div>
+
+                                    <div data-newbie-hidden data-tg-type="ooba" id="xtc_block" class="wide100p">
+                                        <h4 class="wide100p textAlignCenter">
+                                            <label data-i18n="XTC Parameters">XTC Parameters</label>
```
Comment (p-e-w) [inline public/index.html]:
```suggestion
                                            <label data-i18n="Exclude Top Choices (XTC)">Exclude Top Choices (XTC)</label>
```

File: public/index.html [hunk 2]
```diff
@@ -1291,6 +1291,26 @@ <h4 class="wide100p textAlignCenter">
...
+                                            <div class="fa-solid fa-circle-info opacity50p" data-i18n="[title]XTC_Parameters_desc" title="XTC (eXtreme Tail Cutting) parameters for controlling token sampling"></div>
```
Comment (p-e-w) [inline public/index.html]:
```suggestion
                                            <a href="https://github.com/oobabooga/text-generation-webui/pull/6335" target="_blank">
                                                <div class=" fa-solid fa-circle-info opacity50p"></div>
                                            </a>
```

File: public/index.html [hunk 3]
```diff
@@ -1291,6 +1291,26 @@ <h4 class="wide100p textAlignCenter">
...
+                                        </h4>
+                                        <div class="flex-container flexFlowRow gap10px flexShrink">
+                                            <div class="alignitemscenter flex-container flexFlowColumn flexBasis48p flexGrow flexShrink gap0">
+                                                <small data-i18n="XTC Threshold">XTC Threshold</small>
```
Comment (p-e-w) [inline public/index.html]:
```suggestion
                                                <small data-i18n="Threshold">Threshold</small>
```

File: public/index.html [hunk 4]
```diff
@@ -1291,6 +1291,26 @@ <h4 class="wide100p textAlignCenter">
...
+                                                <input class="neo-range-slider" type="range" id="xtc_threshold_textgenerationwebui" name="volume" min="0" max="1" step="0.01" />
+                                                <input class="neo-range-input" type="number" min="0" max="1" step="0.01" data-for="xtc_threshold_textgenerationwebui" id="xtc_threshold_counter_textgenerationwebui">
+                                            </div>
+                                            <div class="alignitemscenter flex-container flexFlowColumn flexBasis48p flexGrow flexShrink gap0">
+                                                <small data-i18n="XTC Probability">XTC Probability</small>
```
Comment (p-e-w) [inline public/index.html]:
```suggestion
                                                <small data-i18n="Probability">Probability</small>
```

**Review by p-e-w — COMMENTED** [review]
Suggested some changes to bring inputs in line with those for DRY.

I really like the alternative name "eXtreme Tail Cutting" proposed in this PR, but unfortunately, XTC cuts the head, not the tail, so it's probably better to just stick with the official name :)

**Comment by vitorfdl** [comment]
> Suggested some changes to bring inputs in line with those for DRY.
> 
> I really like the alternative name "eXtreme Tail Cutting" proposed in this PR, but unfortunately, XTC cuts the head, not the tail, so it's probably better to just stick with the official name :)

Haha, that was filled in by copilot and I didn't review. Sorry about that and thanks for fixing it

**Comment by Wolfsblvt** [comment]
Looks good.
If ooba merges this, we can go ahead and test and merge this too. I'll mark it as "Waiting" for now.

**Review by Cohee1207 — APPROVED** [review]
LGTM

**Comment by p-e-w** [comment]
Please change the maximum of the threshold slider to 0.5. I just did the same thing in the original PR. XTC only takes effect if at least *two* tokens are above the threshold, so values larger than 0.5 do not make sense.

**Comment by 3dfactor** [comment]
KoboldCPP will also have XTC on next release.

**Comment by Wolfsblvt** [comment]
> KoboldCPP will also have XTC on next release.

Depending on how fast ooba merges this (or if they ever), this should likely be a separate PR then, just using the same base controls and naming, so they don't break on merge.

Do you have a link to the KoboldCPP PR, or are willing to do the PR yourself?

**Comment by SabinStargem** [comment]
KoboldCPP now has XTC.  Here are the code commits.

[KoboldCPP - Added XTC sampler](https://github.com/LostRuins/koboldcpp/commit/5bf527a6aec241249793be17e4e3b7a0dbed59b2#comments)

**Comment by Cohee1207** [comment]
I'll open a separate PR to add ooba-specific stuff.

**Pull request merged**